[fx] introduce `__fx_create_arg__` dunder method for controlling custom classes are handled as node args

bradleyhd opened the pull request:
Summary: These changes would allow objects to control how they are handled when they are an argument to a torch.fx call_module node from within their source. Previously, we have been using a custom Tracer with an overridden create_arg() method and branching based on class name to handle args that are unusual (data classes, etc).

Differential Revision: D27976120

suo reviewed (approved):
Nice! This approach feels more Pythonic to me (using special methods to customize behavior through a protocol vs. overriding the behavior the tracer) and increases the composability of custom tracers.

bradleyhd committed:
[fx] introduce `__fx_create_arg__` dunder method for controlling custom classes are handled as node args (#61780)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/61780

These changes would allow objects to control how they are handled when they are an argument to a torch.fx call_module node from within their source. Previously, we have been using a custom Tracer with an overridden create_arg() method and branching based on class name to handle args that are unusual (data classes, etc).

Reviewed By: suo, houseroad

Differential Revision: D27976120

fbshipit-source-id: e10a27f48af9c18626f8b65508cd12660aa56e94
